airflow/providers/amazon/aws/triggers/s3.py

```diff
@@ -0,0 +1,99 @@
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+
+class S3KeyTrigger(BaseTrigger):
+    """
+    S3KeyTrigger is fired as deferred class with params to run the task in trigger worker.
+
+    :param bucket_name: Name of the S3 bucket. Only needed when ``bucket_key``
+        is not provided as a full s3:// url.
+    :param bucket_key:  The key being waited on. Supports full s3:// style url
+        or relative path from root level. When it's specified as a full s3://
+        url, please leave bucket_name as `None`.
+    :param wildcard_match: whether the bucket_key should be interpreted as a
+        Unix wildcard pattern
+    :param aws_conn_id: reference to the s3 connection
+    :param hook_params: params for hook its optional
+    """
+
+    def __init__(
+        self,
+        bucket_name: str,
+        bucket_key: str | list[str],
+        wildcard_match: bool = False,
+        aws_conn_id: str = "aws_default",
+        poke_interval: float = 5.0,
+        should_check_fn: bool = False,
+        **hook_params: Any,
+    ):
+        super().__init__()
+        self.bucket_name = bucket_name
+        self.bucket_key = bucket_key
+        self.wildcard_match = wildcard_match
+        self.aws_conn_id = aws_conn_id
+        self.hook_params = hook_params
+        self.poke_interval = poke_interval
+        self.should_check_fn = should_check_fn
+
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        """Serialize S3KeyTrigger arguments and classpath."""
+        return (
+            "airflow.providers.amazon.aws.triggers.s3.S3KeyTrigger",
+            {
+                "bucket_name": self.bucket_name,
+                "bucket_key": self.bucket_key,
+                "wildcard_match": self.wildcard_match,
+                "aws_conn_id": self.aws_conn_id,
+                "hook_params": self.hook_params,
+                "poke_interval": self.poke_interval,
+                "should_check_fn": self.should_check_fn,
+            },
+        )
+
+    @cached_property
+    def hook(self) -> S3Hook:
+        return S3Hook(aws_conn_id=self.aws_conn_id, verify=self.hook_params.get("verify"))
+
+    async def run(self) -> AsyncIterator[TriggerEvent]:
+        """Make an asynchronous connection using S3HookAsync."""
+        try:
+            async with self.hook.async_conn as client:
+                while True:
+                    if await self.hook.check_key_async(
+                        client, self.bucket_name, self.bucket_key, self.wildcard_match
+                    ):
+                        if self.should_check_fn:
+                            s3_objects = await self.hook.get_files_async(
+                                client, self.bucket_name, self.bucket_key, self.wildcard_match
+                            )
+                            await asyncio.sleep(self.poke_interval)
+                            yield TriggerEvent({"status": "running", "files": s3_objects})
```

Comment:
I don't think this is a good idea. It means that every `poke_interval` you're going to go back to the scheduler, to find a worker, only to run a method and send the job back to the triggerer to sleep one more interval...

Maybe I'm too worried, but it's going to put a lot of pressure on the scheduler, and it's also a lot of back and forth between workers and triggerer, which would make the balance tilt towards not using deferrable in that case.

Or we need to find a way to pass whatever we want to do in that check_fn to the triggerer. Maybe a [JMESPath](https://jmespath.org/) expression ?

Reply:
I think that if `check_fn` is evaluated in the worker it's not deferrable at all. And scheduler would handle the deferring it for `check_fn` python callable every time. JMESPath or any other possible approach can come in as a separate PR for this. I can create a issue for this